### PR TITLE
Allow DQL functions to specify return type

### PR DESCRIPTION
**Pull request by Grafikart**
Some functions have a predefined type and could be resolved correctly by the SQLWalker. I don't want to change things too much but it could be possible to do the same thing for other functions like SUM for instance that would be resolved to float.

This is the follow up of the #7641 PR (I won't delete it this time sorry for the inconvenience)

### CODE REVIEW

File: lib/Doctrine/ORM/Query/AST/Functions/FunctionNode.php
Comment (ostrolucky) [inline lib/Doctrine/ORM/Query/AST/Functions/FunctionNode.php]: So copying my comment here: 

Why defining this when it's default fallback anyway?

**Review by ostrolucky — COMMENTED** [review]
Copying my review here 

It's a nice, simple patch solving the problem, however I'm not sure we should bite the bullet with floats. COUNT() and LENGTH() as integer types are fine, but casting to float in PHP will mean losing accuracy. It's also quite strange to make eg. ABS(1) return 1.0. Also, this a BC break and as such should be documented.

If more dpeople could share their opinion about what I writtent, that would be nice too

File: lib/Doctrine/ORM/Query/AST/Functions/FunctionNode.php
Comment (Grafikart) [reply lib/Doctrine/ORM/Query/AST/Functions/FunctionNode.php]: I copied this here so child classes automatically implement the TypedExpression interface.

**Comment by Grafikart** [comment]
@ostrolucky If you want I could postpone the change on ABS() so the new feature can be merged without BC (knowing we would have to change it in the future).

**Comment by ostrolucky** [comment]
Before doing that, let's see what other doctrine maintainers think. WDYT @doctrine/doctrinecore ?

File: lib/Doctrine/ORM/Query/AST/Functions/FunctionNode.php
Comment (ostrolucky) [reply lib/Doctrine/ORM/Query/AST/Functions/FunctionNode.php]: Why should they automatically implement it though? Then they need to override it, instead of cleanly implement it.

Comment (Grafikart) [reply lib/Doctrine/ORM/Query/AST/Functions/FunctionNode.php]: This is an opinionated decision (I didn't want to specify "implements ...." on subclass). If it really bother I won't argue and fix my code according to your review

**Comment by SenseException** [comment]
master is the next major release. Isn't this a fitting time to introduce a BC break?

File: lib/Doctrine/ORM/Query/AST/Functions/FunctionNode.php
Comment (lcobucci) [reply lib/Doctrine/ORM/Query/AST/Functions/FunctionNode.php]: @Grafikart I understand your decision. However, having the concrete classes implementing the interface makes things much more explicit - especially if you consider that we have many extensions providing platform-specific functions.

Please do apply the changes to have the type only on the concrete classes :+1:

File: lib/Doctrine/ORM/Query/AST/TypedExpression.php
Comment (lcobucci) [inline lib/Doctrine/ORM/Query/AST/TypedExpression.php]: As mentioned in another comment we have extensions that could benefit from having DBAL handling the type conversion, so I think this doesn't necessarily need to be marked as `@internal`.

File: tests/Doctrine/Tests/ORM/Query/SelectSqlMappingTest.php
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Query/SelectSqlMappingTest.php]: I think that we should turn this test into a functional one, avoiding the use of reflection - which makes the test somewhat fragile.

**Review by lcobucci — CHANGES_REQUESTED** [review]
This changeset is quite nice indeed. The title is a bit misleading since the previous behaviour was working as intended, I'd consider this as an improvement of the API.

The usage of floats for SUM(), AVG(), and ABS() might indeed cause some trouble with precision (which is why we don't do casts for decimal columns). I'd say we should leave them as string for now and have this targeting v2.8.0.

File: lib/Doctrine/ORM/Query/AST/TypedExpression.php
Comment (Grafikart) [reply lib/Doctrine/ORM/Query/AST/TypedExpression.php]: This is fixed

File: tests/Doctrine/Tests/ORM/Query/SelectSqlMappingTest.php
Comment (Grafikart) [reply tests/Doctrine/Tests/ORM/Query/SelectSqlMappingTest.php]: I don't know how to solve this test. I can't find another entry point to reach the resultSetMapping to check if everything is working as expected.

Comment (lcobucci) [reply tests/Doctrine/Tests/ORM/Query/SelectSqlMappingTest.php]: Don't worry, we can take care of that. Thanks a lot for your help!

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH7941Test.php
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH7941Test.php]: @Grafikart this is what I meant with a more functional approach :smile: 

The main advantage is that we only rely on the public API here, which makes it easier to detect BC-breaks.

By using `self::assertSame()` we check that both value and type match expected value and `self::assertRegExp()` only accepts strings.

**Review by lcobucci — APPROVED** [review]
This is good to go IMHO, it would be nice to have another pair of eyes here.

It might be good to document a minor behaviour change regarding the return type of `COUNT()` and `LENGTH()` - there's no need for casting it to integer anymore. What do you think @doctrine/team-orm-documentation?

**Review by ostrolucky — APPROVED** [review]

**Comment by lcobucci** [comment]
@Grafikart thanks! :ship:

**Comment by SenseException** [comment]
In some cases functions like `COUNT` are already documented with an integer return type: https://www.doctrine-project.org/projects/doctrine-orm/en/2.7/reference/dql-doctrine-query-language.html#pure-and-mixed-results
This is now some kind of self-fulfilling prophecy (even though the change is for 2.8).

**Comment by VincentLanglet** [comment]
@Grafikart When I use `COUNT` with `getSingleResult`, I now get
```
[ 0 => 42 ]
```
But if I use `getSingleScalarResult` I get `'42'`.

Shouldn't the result be consistent ?

**Pull request merged**
